### PR TITLE
Basic request checking

### PR DESCRIPTION
**Pull request by dachrillz**
Add basic request body checking
We now check the request body to see if content was set. For now only
works with application/json

merge #13 first

### CODE REVIEW

File: scenarios/simple/requestBodyNotDefined/http.json [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  "log": {
+    "entries": [
+      {
+        "request": {
+          "method": "GET",
+          "url": "http://localhost/ping",
+          "postData": {
+            "mimeType": "application/json",
+            "text": "{\"hello\":\"world\"}"
+          }
```
Comment (bittrance) [inline scenarios/simple/requestBodyNotDefined/http.json:11]: GET requests cannot have a body. We do not at this point validate the OpenAPI spec, but expect it to be sane, so we should not have invalid examples.

File: src/api/model.ts [hunk 2]
```diff
@@ -12,7 +12,10 @@ export type HTTPMETHOD =
 export type ApiError =
   | "METHOD NOT FOUND"
   | "PATH NOT FOUND"
-  | "STATUS NOT FOUND";
+  | "STATUS NOT FOUND"
+  | "REQUEST BODY NOT FOUND"
+  | "CONTENT WAS NOT FOUND"
+  | "BAD MIMETYPE";
```
Comment (bittrance) [inline src/api/model.ts]: Reviewing this batch of PRs, I'm realizing that we probably should formulate these in a "neutral" fashion: "The spec says X but the request has Y" sort of messages. I think we should avoid assuming that either side is right.

File: src/api/index.ts [hunk 3]
```diff
@@ -6,6 +6,14 @@ const readApi = async (path: string): Promise<a.Root> => {
   return api;
 };
 
+const getMimeType = (content: a.Content): a.MimeType | undefined => {
+  if (content["application/json"]) {
+    return a.newMimeType({
+      items: Object.keys(content["application/json"].schema).length,
+    });
+  }
+};
```
Comment (bittrance) [inline src/api/index.ts:15]: Note that in the special-case of JSON, `application/json` and `text/json` are synonymous (since JSON prescribes UTF8 and ignores newlines). We can thus support both.

File: src/api/index.ts [hunk 4]
```diff
@@ -53,6 +58,20 @@ const getParameters = (parameters?: a.Parameter[]) => {
   });
 };
 
+const getRequestBody = (
```
Comment (bittrance) [inline src/api/index.ts:61]: I'm not sure I'm completely into the terminology, but isn't this `getRequestBodyResultNode` or something?

File: src/rule/index.ts [hunk 5]
```diff
@@ -120,6 +120,50 @@ export const removeServer = (
...
+    result.success = false;
+    resultOperationNode.requestBody.content = a.newContentBodyWithErrors(
+      "CONTENT WAS NOT FOUND"
+    );
+    return;
+  }
+  incrementHit(operationNode.requestBody.content);
+  resultOperationNode.requestBody.content = a.newContentBody();
+
+  const mimeType = request.postData?.mimeType as string;
```
Comment (bittrance) [inline src/rule/index.ts]: This is `requestMimeType`.

File: src/rule/index.ts [hunk 6]
```diff
@@ -120,6 +120,50 @@ export const removeServer = (
...
+  if (
+    mimeType === "application/json" &&
+    !operationNode.requestBody.content[mimeType]
+  ) {
```
Comment (bittrance) [inline src/rule/index.ts]: Why is this dependent on `application/json`? The MIME type should always match whether we can validate the body?

File: src/rule/index.ts [hunk 7]
```diff
@@ -195,6 +240,10 @@ const verifyRequest = (
     });
   }
 
+  if (request.postData) {
+    matchRequestBody(operationNode, request, resultOperationNode, result);
+  }
```
Comment (bittrance) [inline src/rule/index.ts:250]: This does not cover the case that the spec says the request should have a body but it does not?

**Review by bittrance — COMMENTED** [review]
Please have a look at the comments and see if you agree that there is more to do on this PR.

File: src/api/model.ts (hunk 2 above)
Comment (dachrillz) [reply src/api/model.ts]: agreed, but I feel like we should rewrite the whole error handling once we have the basic features in place.

RIght now I'm not entirerly sure what I actually want reported back to me.

I'm thinking that we could borrow a few ideas from how compilers report errors maybe.

File: scenarios/simple/requestBodyNotDefined/http.json (hunk 1 above)
Comment (dachrillz) [reply scenarios/simple/requestBodyNotDefined/http.json:11]: changed the scenario to use a post

File: src/api/index.ts (hunk 4 above)
Comment (dachrillz) [reply src/api/index.ts:61]: I seem to have named all these get functions to reflect what is retrieved from the api specification tree. 

E.g. in this case its named after https://swagger.io/docs/specification/describing-request-body/

We could change this terminology since I agree that it is a bit confusing. 

I renamed another function that had the name node just to make the naming at least consistent

File: src/rule/index.ts (hunk 5 above)
Comment (dachrillz) [reply src/rule/index.ts]: fixed

File: src/rule/index.ts (hunk 7 above)
Comment (dachrillz) [reply src/rule/index.ts:250]: Added a test and check for this!

File: src/rule/index.ts (hunk 6 above)
Comment (dachrillz) [reply src/rule/index.ts]: this was dues to some type wrangling, made it somewhat better in https://github.com/XenitAB/apidelta/pull/14/commits/f4fff90e8d058e32262067d1e4b9070cc2d7b3ec

File: src/api/index.ts (hunk 3 above)
Comment (dachrillz) [reply src/api/index.ts:15]: Added support for text/json as well. This meant i could remove a lot of if(var === "application/json")

https://github.com/XenitAB/apidelta/pull/14/commits/f4fff90e8d058e32262067d1e4b9070cc2d7b3ec

File: src/api/api.test.ts [hunk 8]
```diff
@@ -23,19 +23,11 @@ const expected: Root = {
...
-                        type: "string",
-                        example: "doggie",
-                      },
-                    }),
-                  }),
-                }),
-              },
+              "application/json": expect.objectContaining({
+                x_x_x_x_results: {
+                  hits: 0,
```
Comment (bittrance) [inline src/api/api.test.ts:28]: I should perhaps have caught this before, but shouldn't it start at 1? That may be a separate PR tho.

File: src/api/model.ts [hunk 9]
```diff
@@ -122,15 +126,44 @@ export type Parameter = {
...
+    x_x_x_x_results: newReportItemWithError(error),
+  };
+};
+
+export type Mime = "application/json" | "text/json";
+
 export type Content = {
-  "application/json"?: any;
+  "application/json"?: MimeType;
+  "text/json"?: MimeType;
```
Comment (bittrance) [inline src/api/model.ts:166]: This looks a bit strange given that we know that these two will be identical.

**Review by bittrance — APPROVED** [review]
I think "start at zero" comment is still relevant, but other than that, this looks fine.
